File: 0986-interval-list-intersections/0986-interval-list-intersections.java
class Solution {
    public int[][] intervalIntersection(int[][] firstList, int[][] secondList) {
         int i=0,j=0;
        List<int[]> result= new ArrayList<>();
        while(i<firstList.length&&j<secondList.length){
            int[] first=firstList[i];
            int[] second=secondList[j];
            if((first[0]>=second[0]&&first[0]<=second[1])||(second[0]>=first[0]&&second[0]<=first[1])){
                result.add(new int[]{Math.max(first[0],second[0]),Math.min(first[1],second[1])});
            }
            if(first[1]<second[1]){
                i++;
            }else{
                j++;
            }
        }
        return result.toArray(new int[result.size()][]);
    }
}

File: 0986-interval-list-intersections/0986-interval-list-intersections.py
class Solution:
    def intervalIntersection(self, A: List[List[int]], B: List[List[int]]) -> List[List[int]]:
        result = []
        i, j, start, end = 0, 0, 0, 1
        

        while i < len(A) and j < len(B):
            A_O_B = A[i][0] >= B[j][0] and A[i][0] <= B[j][1]
            B_O_A = B[j][0] >= A[i][0] and B[j][0] <= A[i][1]
            if (A_O_B or B_O_A):
                result.append([max(A[i][0], B[j][0]), min(A[i][1], B[j][1])])
            if A[i][1] < B[j][1]:
                i += 1
            else:
                j += 1

        return result
        